File: CoreUI/Sources/CoreUI/Design/Typography/Typography.swift
import SwiftUI

public enum Typography {

    public static let largeTitle = Style(size: 32, weight: .bold, tracking: -0.8, color: ColorPalette.ink)
    public static let title = Style(size: 22, weight: .heavy, tracking: -0.6, color: ColorPalette.ink)
    public static let headline = Style(size: 19, weight: .bold, color: ColorPalette.ink)
    public static let body = Style(size: 17, weight: .semibold, color: ColorPalette.ink)
    public static let calloutBold = Style(size: 15, weight: .semibold, color: ColorPalette.ink)
    public static let buttonLabel = Style(size: 15, weight: .semibold, tracking: 0.6, color: ColorPalette.inkInverse)
    public static let callout = Style(size: 15, weight: .regular, color: ColorPalette.textSecondary)
    public static let footnote = Style(size: 13, weight: .regular, color: ColorPalette.textSecondary)
    public static let label = Style(size: 12, weight: .semibold, tracking: 1.4, color: ColorPalette.textTertiary)
    public static let micro = Style(size: 10, weight: .semibold, tracking: 1.8, color: ColorPalette.accent)

}

public extension Typography {

    struct Style {

        public let font: Font
        public let tracking: CGFloat
        public let color: Color

        init(size: CGFloat, weight: Font.Weight, tracking: CGFloat = 0, color: Color) {
            font = .system(size: size, weight: weight)
            self.tracking = tracking
            self.color = color
        }

    }

}

File: androidApp/core-ui/src/main/kotlin/com/zeitnot/android/coreui/Typography.kt
package com.zeitnot.android.coreui

import androidx.compose.ui.graphics.Color
import androidx.compose.ui.text.TextStyle
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.sp

object Typography {

    val clockDigits = style(
        size = 76,
        weight = FontWeight.Bold,
        tracking = -2.3f,
        hasTabularFigures = true,
        color = ColorPalette.ink)

    val largeTitle = style(size = 32, weight = FontWeight.Bold, tracking = -0.8f, color = ColorPalette.ink)
    val title = style(size = 22, weight = FontWeight.ExtraBold, tracking = -0.6f, color = ColorPalette.ink)
    val headline = style(size = 19, weight = FontWeight.Bold, color = ColorPalette.ink)
    val body = style(size = 17, weight = FontWeight.SemiBold, color = ColorPalette.ink)
    val calloutBold = style(size = 15, weight = FontWeight.SemiBold, color = ColorPalette.ink)
    val buttonLabel = style(size = 15, weight = FontWeight.SemiBold, tracking = 0.6f, color = ColorPalette.inkInverse)
    val callout = style(size = 15, weight = FontWeight.Normal, color = ColorPalette.textSecondary)
    val footnote = style(size = 13, weight = FontWeight.Normal, color = ColorPalette.textSecondary)
    val label = style(size = 12, weight = FontWeight.SemiBold, tracking = 1.4f, color = ColorPalette.textTertiary)
    val playerName = style(size = 12, weight = FontWeight.Bold, tracking = 2.4f, color = ColorPalette.textSecondary)
    val micro = style(size = 10, weight = FontWeight.SemiBold, tracking = 1.8f, color = ColorPalette.accent)

    private fun style(
        size: Int,
        weight: FontWeight,
        tracking: Float = 0f,
        hasTabularFigures: Boolean = false,
        color: Color
    ) = TextStyle(
        fontSize = size.sp,
        fontWeight = weight,
        letterSpacing = tracking.sp,
        fontFeatureSettings = if (hasTabularFigures) TABULAR_FIGURES else null,
        color = color)

}

private const val TABULAR_FIGURES = "tnum"
